### PR TITLE
Crash fixed when no tag is provided for the container image

### LINKED ISSUE #39: Crash if no tag is provided for the container image
The agent crashes if no tag is provided for the container image.

Given the following `kafka.yaml`:
```yaml
apiVersion: v1
kind: Pod
metadata:
  name: kafka
spec:
  containers:
  - name: kafka
    image: kafka
  tolerations:
  - key: "kubernetes.io/arch"
    operator: "Equal"
    value: "stackable-linux"
```

When the pod is deployed:
```sh
$ kubectl apply -f kafka.yml
```

Then the agent crashes with the following log output:
```
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', src/provider/repository/package.rs:42:35
```

### PR DESCRIPTION
**Pull request by siegfriedweber**
Fixes #39 

When no tag is provided for the container image then the agent crashes:

```
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', src/provider/repository/package.rs:42:35
```

With this change the pod is rejected with an according log message:

```
2021-01-20T14:18:58Z WARN  kubelet::kubelet] Error resyncing pods: Invalid content in pod object: Unable to get package reference from pod "kafka": Tag is required.
```

### CODE REVIEW

**Review by soenkeliebau — DISMISSED** [review]
Looks very good to me!

**Comment by soenkeliebau** [comment]
I spoke with Lars this morning and he will try to review this PR this afternoon.

File: src/provider/mod.rs [hunk 1]
```diff
@@ -87,23 +88,23 @@ impl StackableProvider {
...
-            };
-            Err(e)
+        if let Some((container, [])) = pod.containers().split_first() {
+            container
+                .image()
+                .and_then(|maybe_ref| maybe_ref.ok_or(anyhow!("Image is required.")))
+                .and_then(Package::try_from)
+                .map_err(|err| PodValidationError {
+                    msg: format!(
+                        r#"Unable to get package reference from pod "{}": {}"#,
```
Comment (lfrancke) [inline src/provider/mod.rs:98]: We have a kinda looseish convention of wrapping values in square brackets:

```suggestion
                        "Unable to get package reference from pod [{}]: {}",
```

File: src/provider/mod.rs [hunk 2]
```diff
@@ -206,3 +207,89 @@ impl Provider for StackableProvider {
...
+
+        if let Ok(package) = maybe_package {
+            assert_eq!("kafka", package.product);
+            assert_eq!("2.7", package.version);
+        } else {
+            panic!("Package expected but got {:?}", maybe_package);
+        }
+    }
+
+    #[rstest(pod_config, expected_err,
```
Comment (lfrancke) [inline src/provider/mod.rs:240]: This is fabulous!

(should you update the string on top you need to update the fixtures as well)

File: Cargo.toml [hunk 3]
```diff
@@ -30,3 +30,8 @@ thiserror = "1.0"
 url = "2.2"
 pnet = "0.26.0"
 stackable_config = { git = "https://github.com/stackabletech/common.git", branch = "main" }
+
+[dev-dependencies]
+indoc = "1.0"
```
Comment (lfrancke) [inline Cargo.toml:35]: TIL about rstest and indoc. Nice!

**Review by lfrancke — CHANGES_REQUESTED** [review]
Only minor stuff.
This looks fabulous.
If this is your first Rust PR I'm depressed :) Good job!

File: src/provider/repository/package.rs [hunk 4]
```diff
@@ -31,15 +30,18 @@ impl Package {
...
-    type Error = StackableError;
+    type Error = anyhow::Error;
 
     // Converts from an oci reference to a package representation
     // The oci tag (anything after the \":\" in the string) is used as
     // version by this code and needs to be present
-    fn try_from(value: Reference) -> Result<Self, Self::Error> {
+    fn try_from(value: Reference) -> Result<Self> {
+        let repository = value.repository();
+        let tag = value.tag().ok_or(anyhow!("Tag is required."))?;
```
[the remark this answers is not in the document]
Comment (siegfriedweber) [reply src/provider/repository/package.rs]: Clippy is right. Branches in the control flow should be evaluated lazily. I fixed this.

File: src/provider/mod.rs [hunk 5]
```diff
@@ -87,23 +88,23 @@ impl StackableProvider {
...
-        let containers = pod.containers();
-        return if containers.len().ne(&1) {
-            let e = PodValidationError {
-                msg: String::from("Size of containers list in PodSpec has to be exactly 1"),
-            };
-            Err(e)
+        if let Some((container, [])) = pod.containers().split_first() {
+            container
+                .image()
+                .and_then(|maybe_ref| maybe_ref.ok_or(anyhow!("Image is required.")))
```
[the remark this answers is not in the document]
Comment (siegfriedweber) [reply src/provider/mod.rs]: Fixed according to clippy's suggestion.

File: src/provider/mod.rs (hunk 1 above)
Comment (siegfriedweber) [reply src/provider/mod.rs]: Convention for error messages applied

File: src/provider/mod.rs (hunk 2 above)
Comment (siegfriedweber) [reply src/provider/mod.rs:240]: Fixtures updated

File: Cargo.toml (hunk 3 above)
Comment (siegfriedweber) [reply Cargo.toml:35]: They seem to be very popular.

**Review by lfrancke — APPROVED** [review]
Looks good to me, thank you!

**Pull request merged**